S3 Backup design doc (#2627)
dandi/dandi-archive: doc/design/s3-backup.md

- **Glacier Deep Archive requires 12-48 hours to retrieve data:** the exact amount of time depends on the retrieval mechanism (see [AWS docs](https://aws.amazon.com/s3/storage-classes/glacier/)), but either way, instantaneous recovery is not possible. However, this latency is acceptable for purposes of catastrophic data recovery; it also indicates why the trailing delete mechanism continues to be valuable (it offers a much faster "data recovery" in the cases where it can be used).

## Cost

**Below are the additional costs introduced by this backup feature** for a 1 PB primary bucket (assuming both the primary bucket and backup bucket are in us-east-2). All of this information was gathered from the "Storage & requests", "Data transfer", and "Replication" tabs on [https://aws.amazon.com/s3/pricing/](https://aws.amazon.com/s3/pricing/).

**Storage Costs** (backup bucket in us-east-2):

- Glacier Deep Archive storage: ~$0.00099/GB/month
    - 1 PB = 1,024 TB × $0.99/TB = **$1,014/month**

**Data Transfer Costs**:

> - Same-region data transfer between S3 buckets is free

Reviewer: The definition on line https://github.com/dandi/dandi-archive/pull/2627/files#diff-9a9b9bb814eeea229ca42a1f536abf895736f6aeb248def55cc122469f619d0fR6 indicates the backup will live in a separate region though?

Author: Ah thanks, I missed that - we originally considered a bucket in a separate region, but settled on same-region. I pushed a fix

**Retrieval Costs** (only incurred when disaster recovery is needed):

- Glacier Deep Archive retrieval:
    - $0.02/GB (standard, 12-hour retrieval)

>     - $0.025/GB (bulk retrieval, can take up to 48 hours)

Reviewer: Can you produce worst-case full bucket retrieval summary estimates here? Similar to https://github.com/dandi/dandi-archive/pull/2627/files#diff-9a9b9bb814eeea229ca42a1f536abf895736f6aeb248def55cc122469f619d0fR60

Reviewer: Since that cost is very important to consider to get full use from this endeavor

### Future Costs

> The DANDI Archive is expecting a ramp-up in data volume of 6 PB of new data over each of the next five years, culminating in a total of 30PB.

Reviewer 2: > I thought it was 1 PB per year for a total of 6 PB by the end? And then a separate potential project that was also several PB

I may have misunderstood. @kabilar probably knows the answer.

> Also we didn't get 100% confirmation of S3 bucket size extension yet either did we? (guessing it is fine for first 1-2 years up to 2-3 PB total storage)

I was not privy to those conversations. This one is for @satra.

> Scaling up the previous analysis means that the monthly costs will be projected to rise to a total of **~$31,000/month** once all of that data is seated. While $1000/month may be a feasible ongoing cost, $30000/month is not.

Reviewer: The output of `df -hT .`

```
Filesystem                   Type
hstor004-n1:/group/dandi/001    nfs4
```

## Requirements

### Functional requirements

- **Replication.** Maintain a copy of the current contents of the primary bucket in the backup bucket.
- **Backup.** Replicate all primary bucket writes and deletes to the backup bucket, maintaining a "tape archive" style history-based backup.

### Non-functional requirements

- **Cost control.** Enable tunable controls to help keep costs reasonable for the expected data volume.

## Proposed Solution

The solution leverages the following native S3 features to implement automatic replication with history and optional delayed deletion:

1. **S3 Replication** to copy objects from the primary bucket to a backup bucket.
2. **Delete Marker Replication** to propagate soft deletions to the backup.
3. **Bucket Versioning** on both buckets to create a history-based backup and support the replication and recovery mechanisms

> 4. **Optional Lifecycle Policies** on the backup bucket to permanently delete objects after a retention period

Reviewer: ```suggestion
3. **Bucket Versioning** on both buckets to create a history-based backup and support the replication and recovery mechanisms.
4. **Optional Lifecycle Policies** on the backup bucket to permanently delete objects after a retention period.
```

for consistency (the above lines end in periods)

    - $0.0025/GB (bulk retrieval, can take up to 48 hours)

Imagining that the entire primary bucket was destroyed (which is not the
expected scale of data loss, but useful as a worst-case analysis), then the cost
to restore from backup would be
$`1\ \rm{PB} \times 1024\ \rm{TB}/\rm{PB} \times 1024\rm{GB}/\rm{TB} \times \$0.0025/GB/mo \approxeq \$2600`$.

### Future Costs

The DANDI Archive is expecting a ramp-up in data volume of 1 PB of new data over each of the next five years, culminating in a total of 6PB.

Scaling up the previous analysis means that the monthly costs will be projected to rise to a total of **~$6,100/month** once all of that data is seated.
The worst-case disaster recovery cost would similarly scale up to a total of **~$16000**.

An open question is whether the AWS Open Data Sponsorship program would cover the marginal costs of backup. A quick estimate shows that once all 5 PB has been uploaded, the expected bucket cost for the primary bucket (i.e., what the AWS Open Data Sponsorship program covers already, excluding backup) will be:

$$
6\ \rm{PB} \times 1024\ \rm{TB}/\rm{PB} \times 1024\ \rm{GB}/\rm{TB} \times \$0.021/GB/mo \approxeq \$130000/mo

> $$

Reviewer: Recommend adjusting as a I showed above

Reviewer 2: 6b69903917d539ef3aa17558b31a16f5788d4dbf

> while the associated backup costs would represent only $`\$31000 / \$660000 \approxeq 4.6\%`$ of the cost of the storage itself.

Reviewer 2: b3e0a9d

to restore from backup would be

$`1\ \rm{PB} \times \frac{1000 \ \rm{TB}}{\rm{PB}} \times \frac{1000 \ \rm{GB}}{\rm{TB}} \times \frac{\$0.0025/mo}{\rm{GB}} = \$2500`$.

### Future Costs

The DANDI Archive is expecting a ramp-up in data volume of 1 PB of new data over each of the next five years, culminating in a total of 6PB.

Scaling up the previous analysis means that the monthly costs will be projected to rise to a total of **~$5,900/month** once all of that data is seated.
The worst-case disaster recovery cost would similarly scale up to a total of **~$16,000**.

An open question is whether the AWS Open Data Sponsorship program would cover the marginal costs of backup. A quick estimate shows that once all 5 PB has been uploaded, the expected bucket cost for the primary bucket (i.e., what the AWS Open Data Sponsorship program covers already, excluding backup) will be:

$$
6\ \rm{PB} \times \frac{1000\ \rm{TB}}{\rm{PB}} \times \frac{1000\ \rm{GB}}{\rm{TB}} \times \frac{\$0.021/mo}{\rm{GB}} \approxeq \$126000/mo
$$

while the associated backup costs would represent only an additional $`\$5900 / \$126000 \approxeq 4.6\%`$ of the cost of the storage itself.

> To help provide a significant level of safety to an important dataset, AWS may be willing to cover such a low marginal cost.

Reviewer: ```suggestion
To help provide a significant level of safety to an important database, it may be worth reaching out to see if AWS may be willing to cover such a low marginal cost.
```

Original wording sounds as if we are speaking for AWS

Reviewer: Please remove this line as it has been confirmed that the Open Data program will not cover backup.

Reviewer 3: ```suggestion
```

> To help provide a significant level of safety to an important dataset, AWS may be willing to cover such a low marginal cost.

Reviewer 4: @satra The design here looks good, but the main open question and blocker is cost.

Would the AWS Open Data program be open to covering the back up storage costs as well (~$12,000/year for a 1 PB backup, and ~$72,000/year for the projected 6 PB backup)? See the doc for more details.
